### PR TITLE
Adds more convolution operations and dynamic op improvments

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

Please review
https://github.com/deeplearning4j/deeplearning4j/blob/master/CONTRIBUTING.md before opening a pull request.

### CODE REVIEW

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunction.java [hunk 1]
```diff
@@ -695,5 +704,10 @@ protected void validateDifferentialFunctionsameDiff(
 
     }
 
+    protected int fromBoolean(boolean bool) {
```
Comment (RobAltena) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunction.java]: This function probably belongs somewhere in the [nd4j utils](https://github.com/deeplearning4j/nd4j/tree/a245a89a9db64272d330a3091c22e3171d5fe1a4/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/util).

Comment (RobAltena) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunction.java]: Or just inline it (which the compiler will probably do anyway).

Comment (agibsonccc) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunction.java]: Yeah agreed. I was messing with how the int/boolean arguments were working.  Retyping that over and over again got annoying :P. I'm shocked this isn't in guava or something (I looked :()

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/DeConv2D.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import org.nd4j.linalg.api.ops.BaseTransformOp;
+import org.nd4j.linalg.api.ops.DynamicCustomOp;
+import org.nd4j.linalg.api.ops.Op;
+import org.nd4j.linalg.factory.Nd4j;
+
+import java.util.List;
+
+
+/**
+ * Pooling2DDerivative operation
```
Comment (maxpumperla) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/DeConv2D.java]: Deconv2D operation

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/Conv3DDerivative.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import org.nd4j.autodiff.ArrayField;
+import org.nd4j.autodiff.functions.DifferentialFunction;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.List;
+
+
+/**
+ * Pooling2DDerivative operation
```
Comment (maxpumperla) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/Conv3DDerivative.java]: Conv3DDerivative

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/FullConv3DDerivative.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.nd4j.autodiff.ArrayField;
+import org.nd4j.autodiff.functions.DifferentialFunction;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.List;
+
+
+/**
+ * Pooling2DDerivative operation
```
Comment (maxpumperla) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/FullConv3DDerivative.java]: FullConv3DDerivative

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/LocalResponseNormalizationDerivative.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import org.nd4j.autodiff.ArrayField;
+import org.nd4j.autodiff.functions.DifferentialFunction;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.List;
+
+
+/**
+ * Pooling2DDerivative operation
```
Comment (maxpumperla) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/LocalResponseNormalizationDerivative.java]: LocalResponseNormalizationDerivative operation

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/SConv2D.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import org.nd4j.autodiff.ArrayField;
+import org.nd4j.autodiff.functions.DifferentialFunction;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.List;
+
+
+/**
+ * Pooling2DDerivative operation
```
Comment (maxpumperla) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/SConv2D.java]: SConv2D

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/SConv2DDerivative.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.nd4j.autodiff.ArrayField;
+import org.nd4j.autodiff.functions.DifferentialFunction;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.List;
+
+
+/**
+ * Pooling2DDerivative operation
```
Comment (maxpumperla) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/convolution/SConv2DDerivative.java]: SConv2DDerivative

**Comment by agibsonccc** [comment]
@maxpumperla yeah thanks missed that..will clean the names up. Copy pasta :D.

**Comment by maxpumperla** [comment]
@agibsonccc yeah, happens all the time. :) Just wanted to point out all the doc string that need updates, so you don't have to (again)

**Pull request merged**